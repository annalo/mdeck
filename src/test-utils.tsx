/* eslint-disable import/no-extraneous-dependencies */

import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/Home/ThemeProvider";
import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <SlideObserverProvider>
          <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
        </SlideObserverProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
