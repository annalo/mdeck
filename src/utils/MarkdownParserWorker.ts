import Worker from "worker-loader!./Worker"; // eslint-disable-line

import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

import { MarkdownContextReducerActionType } from "types/markdown-context";

export default class MarkdownParserWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({ data }: { data: HtmlArray }) =>
      dispatch({
        type: MarkdownContextReducerActionType.SetHtmlArray,
        htmlArray: data,
      });
  }

  parse(md: MarkdownString): void {
    this.worker.postMessage(md);
  }

  terminate(): void {
    if (this.worker) this.worker.terminate();
  }
}
