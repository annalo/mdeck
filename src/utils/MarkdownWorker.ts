import Worker from "worker-loader!./Worker"; // eslint-disable-line
import type { Dispatch } from "react";
import type { MarkdownContextReducerAction } from "types/markdown-context";

export class MarkdownWorker {
  private readonly worker: Worker;

  constructor(dispatch: Dispatch<MarkdownContextReducerAction>) {
    this.worker = new Worker();
    this.worker.onmessage = ({
      data,
    }: {
      data: MarkdownContextReducerAction;
    }) => {
      console.log("markdown worker on message", data);
      dispatch(data);
    };
  }

  parse(md: string): void {
    this.worker.postMessage({ action: "PARSE", md });
  }

  terminate() {
    if (this.worker) this.worker.terminate();
  }
}
