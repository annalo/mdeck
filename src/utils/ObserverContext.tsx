import { createContext } from "react";

export const ObserverContext = createContext<IntersectionObserver | null>(null);
