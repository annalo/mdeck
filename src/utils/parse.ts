import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";
import { injectLineNumber } from "utils/parsePlugins/injectLineNumber";

export const parse = (markdown: string): HtmlArray => {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: HtmlArray } = marpit.render(markdown, {
    htmlAsArray: true,
  });
  return html;
};
