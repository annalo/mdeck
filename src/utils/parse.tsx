/* eslint-disable */
import { Marpit } from "@marp-team/marpit";
import TaskList from "markdown-it-task-lists";

const RULES = [
  "blockquote_open",
  "code_block",
  "fence",
  "heading_open",
  "image",
  "link_open",
  "list_item_open",
  "paragraph_open",
  "table_open",
];

function injectLineNumber(md: any) {
  const { marpit_inline_svg_open } = md.renderer.rules;

  // Enable line sync by per slides
  md.renderer.rules.marpit_inline_svg_open = (tokens, idx, opts, env, self) => {
    const slide = tokens
      .slice(idx + 1)
      .find((t) => t.type === "marpit_slide_open");

    if (slide.map?.length) {
      tokens[idx].attrJoin("class", "src-line");
      tokens[idx].attrSet("data-line", slide.map[0]);
    }

    const renderer = marpit_inline_svg_open || self.renderToken;
    return renderer.call(self, tokens, idx, opts, env, self);
  };

  // Enables line sync per elements
  for (const rule of RULES) {
    const original = md.renderer.rules[rule];

    md.renderer.rules[rule] = (tokens, idx, options, env, self) => {
      const token = tokens[idx];

      if (token.map?.length) {
        token.attrJoin("class", "code-line");
        token.attrSet("data-line", token.map[0]);
      }

      const renderer = original || self.renderToken;
      return renderer.call(self, tokens, idx, options, env, self);
    };
  }
}

export function parse(markdown: string): string {
  const marpit = new Marpit({
    inlineSVG: true,
    markdown: { html: true, linkify: true, typographer: true },
  })
    .use(TaskList)
    .use(injectLineNumber);

  const { html }: { html: string } = marpit.render(markdown);
  return html;
}
