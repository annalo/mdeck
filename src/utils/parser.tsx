import { Marpit } from "@marp-team/marpit";

export function convertMarkdown(markdown: string): string {
  const marpit = new Marpit();
  const { html } = marpit.render(markdown);
  return html;
}
