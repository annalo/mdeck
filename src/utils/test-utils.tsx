/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { render } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { ThemeProvider } from "components/App/ThemeProvider";

import { MarkdownContextProvider } from "contexts/MarkdownContext";
import { PresentationContextProvider } from "contexts/PresentationContext";
import { SlideObserverProvider } from "contexts/SlideObserver";
import { CodeLineObserverProvider } from "contexts/CodeLineObserver";

const AllTheProviders = ({ children }) => {
  return (
    <ThemeProvider>
      <MarkdownContextProvider>
        <PresentationContextProvider>
          <SlideObserverProvider>
            <CodeLineObserverProvider>{children}</CodeLineObserverProvider>
          </SlideObserverProvider>
        </PresentationContextProvider>
      </MarkdownContextProvider>
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

// override render and renderHook methods
export { customRender as render, customRenderHook as renderHook };
