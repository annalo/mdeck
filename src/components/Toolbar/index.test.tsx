import React from "react";
import { render } from "utils/test-utils";
import { Toolbar } from ".";

describe("<Toolbar />", () => {
  test("should render and match the snapshot", () => {
    const { asFragment } = render(<Toolbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
