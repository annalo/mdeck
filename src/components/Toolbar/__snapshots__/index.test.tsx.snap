// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar /> should render and match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  background-color: #455954;
  box-shadow: 0px 0px 1.5px #455954;
  height: 25px;
  overflow: hidden;
  position: fixed;
  bottom: 0;
  width: 100%;
  padding: 0 5px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 25px;
  margin: 0;
  padding: 0;
}

.c3 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 3px;
}

.c6 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 3px;
  opacity: 0.4;
  pointer-events: none;
  margin-left: auto;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 svg {
  -webkit-transition: 0.4s ease-in-out;
  transition: 0.4s ease-in-out;
}

.c7 svg circle,
.c7 svg path {
  stroke: #b5c2c7;
}

.c7 svg:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c7 svg {
  -webkit-transition: 0.4s ease-in-out;
  transition: 0.4s ease-in-out;
}

.c7 svg:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c7 svg:hover {
  fill: #b5c2c7;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 svg {
  -webkit-transition: 0.4s ease-in-out;
  transition: 0.4s ease-in-out;
}

.c4 svg circle,
.c4 svg path {
  stroke: #b5c2c7;
}

.c4 svg:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c4 svg {
  -webkit-transition: 0.4s ease-in-out;
  transition: 0.4s ease-in-out;
}

.c4 svg:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c4 svg:hover {
  fill: #b5c2c7;
}

.c4:hover svg {
  -webkit-transition: opacity 0.4s ease-in-out;
  transition: opacity 0.4s ease-in-out;
  opacity: 0.5;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.c5 {
  position: absolute;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  color: #b5c2c7;
  font-weight: bolder;
  font-size: 8px;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c5:hover {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c2 {
  display: none;
}

<nav
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        accept=".md,.txt"
        class="c2"
        id="fileLoader"
        title="Load Markdown File"
        type="file"
      />
      <li
        aria-label="load-file"
        class="c3"
        role="button"
      >
        <div
          class="c4"
        >
          <svgr-mock />
          <div
            class="c5"
          >
            LOAD
          </div>
        </div>
      </li>
      <li
        aria-label="save-file"
        class="c3"
        role="button"
      >
        <div
          class="c4"
        >
          <svgr-mock />
          <div
            class="c5"
          >
            SAVE
          </div>
        </div>
      </li>
      <li
        aria-label="request-presentation"
        class="c6"
        disabled=""
        role="button"
      >
        <div
          class="c7"
        >
          <svgr-mock />
        </div>
      </li>
    </div>
  </nav>
</DocumentFragment>
`;
